File: instrumentation/restlet/restlet-2.0/library/src/main/java/io/opentelemetry/instrumentation/restlet/v2_0/internal/Experimental.java
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.instrumentation.restlet.v2_0.internal;

import static java.util.logging.Level.FINE;

import io.opentelemetry.instrumentation.restlet.v2_0.RestletTelemetryBuilder;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.logging.Logger;
import javax.annotation.Nullable;

/**
 * This class is internal and experimental. Its APIs are unstable and can change at any time. Its
 * APIs (or a version of them) may be promoted to the public stable API in the future, but no
 * guarantees are made.
 */
public class Experimental {

  private static final Logger logger = Logger.getLogger(Experimental.class.getName());

  @Nullable
  private static final Method emitExperimentalTelemetryMethod =
      getEmitExperimentalTelemetryMethod();

  public void setEmitExperimentalTelemetry(
      RestletTelemetryBuilder builder, boolean emitExperimentalTelemetry) {

    if (emitExperimentalTelemetryMethod != null) {
      try {
        emitExperimentalTelemetryMethod.invoke(builder, emitExperimentalTelemetry);
      } catch (IllegalAccessException | InvocationTargetException e) {
        logger.log(FINE, e.getMessage(), e);
      }
    }
  }

  @Nullable
  private static Method getEmitExperimentalTelemetryMethod() {
    try {
      Method method =
          RestletTelemetryBuilder.class.getDeclaredMethod(
              "setEmitExperimentalHttpServerMetrics", boolean.class);
      method.setAccessible(true);
      return method;
    } catch (NoSuchMethodException e) {
      logger.log(FINE, e.getMessage(), e);
      return null;
    }
  }
}

File: instrumentation/ktor/ktor-2-common/library/src/main/kotlin/io/opentelemetry/instrumentation/ktor/client/internal/Experimental.kt
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.instrumentation.ktor.client.internal

import io.opentelemetry.instrumentation.ktor.client.AbstractKtorClientTelemetryBuilder
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method
import java.util.logging.Level
import java.util.logging.Logger

/**
 * This class is internal and experimental. Its APIs are unstable and can change at any time. Its
 * APIs (or a version of them) may be promoted to the public stable API in the future, but no
 * guarantees are made.
 */
class Experimental {

  companion object {
    private val logger: Logger = Logger.getLogger(Experimental::class.java.name)

    private val emitExperimentalTelemetryMethod = getEmitExperimentalTelemetry()

    fun setEmitExperimentalTelemetry(builder: AbstractKtorClientTelemetryBuilder?, emitExperimentalTelemetry: Boolean) {
      if (emitExperimentalTelemetryMethod != null) {
        try {
          emitExperimentalTelemetryMethod.invoke(builder, emitExperimentalTelemetry)
        } catch (e: IllegalAccessException) {
          logger.log(Level.FINE, e.message, e)
        } catch (e: InvocationTargetException) {
          logger.log(Level.FINE, e.message, e)
        }
      }
    }

    private fun getEmitExperimentalTelemetry(): Method? {
      try {
        val method = AbstractKtorClientTelemetryBuilder::class.java.getDeclaredMethod(
          "setEmitExperimentalHttpClientMetrics",
          Boolean::class.javaPrimitiveType
        )
        method.setAccessible(true)
        return method
      } catch (e: NoSuchMethodException) {
        logger.log(Level.FINE, e.message, e)
        return null
      }
    }
  }
}
